File: src/java/magmac/app/lang/node/StructureMembers.java
package magmac.app.lang.node;

/**
 * Deserializes members inside a class or interface body such as methods and
 * nested structures.
 */

import magmac.api.collect.list.Lists;
import magmac.app.compile.error.CompileResult;
import magmac.app.compile.node.Node;
import magmac.app.compile.rule.OrRule;
import magmac.app.compile.rule.Rule;
import magmac.app.compile.rule.TypeRule;
import magmac.app.lang.Deserializers;
import magmac.app.lang.java.JavaDeserializers;
import magmac.app.lang.JavaRules;
import magmac.app.lang.LazyRule;
import magmac.app.lang.MutableLazyRule;
import magmac.app.lang.java.JavaEnumValues;
import magmac.app.lang.java.JavaLang;
import magmac.app.lang.java.JavaMethod;
import magmac.app.lang.java.JavaStructureMember;
import magmac.app.lang.java.JavaStructureStatement;

public final class StructureMembers {
    public static CompileResult<JavaStructureMember> deserialize(Node node) {
        return Deserializers.orError("structure-members", node, Lists.of(
                Deserializers.wrap(JavaDeserializers::deserializeWhitespace),
                Deserializers.wrap(JavaMethod::deserialize),
                Deserializers.wrap(JavaStructureStatement::deserialize),
                Deserializers.wrap(JavaEnumValues::deserialize),
                Deserializers.wrap(new JavaLang.JavaStructureNodeDeserializer(JavaLang.JavaStructureType.Class)),
                Deserializers.wrap(new JavaLang.JavaStructureNodeDeserializer(JavaLang.JavaStructureType.Interface)),
                Deserializers.wrap(new JavaLang.JavaStructureNodeDeserializer(JavaLang.JavaStructureType.Record)),
                Deserializers.wrap(new JavaLang.JavaStructureNodeDeserializer(JavaLang.JavaStructureType.Enum))
        ));
    }

    public static Rule createClassMemberRule() {
        LazyRule classMemberRule = new MutableLazyRule();
        LazyRule functionSegmentRule = new MutableLazyRule();
        LazyRule valueLazy = new MutableLazyRule();
        var value = JavaRules.initValueRule(functionSegmentRule, valueLazy, "->", JavaRules.createDefinitionRule());
        var functionSegment = FunctionSegments.initFunctionSegmentRule(functionSegmentRule, value, JavaRules.createDefinitionRule());
        return classMemberRule.set(new OrRule(Lists.of(
                JavaRules.createTypedWhitespaceRule(),
                JavaStructureStatement.createStructureStatementRule(new TypeRule("definition", JavaRules.createDefinitionRule()), value),
                JavaMethod.createMethodRule(functionSegment),
                JavaEnumValues.createEnumValuesRule(value),
                JavaRules.createStructureRule("record", classMemberRule),
                JavaRules.createStructureRule("interface", classMemberRule),
                JavaRules.createStructureRule("class", classMemberRule),
                JavaRules.createStructureRule("enum", classMemberRule)
        )));
    }
}

File: src/web/magmac/app/lang/node/StructureMembers.ts
import { Lists } from "../../../../magmac/api/collect/list/Lists";
import { CompileResult } from "../../../../magmac/app/compile/error/CompileResult";
import { Node } from "../../../../magmac/app/compile/node/Node";
import { OrRule } from "../../../../magmac/app/compile/rule/OrRule";
import { Rule } from "../../../../magmac/app/compile/rule/Rule";
import { TypeRule } from "../../../../magmac/app/compile/rule/TypeRule";
import { Deserializers } from "../../../../magmac/app/lang/Deserializers";
import { JavaDeserializers } from "../../../../magmac/app/lang/java/JavaDeserializers";
import { JavaRules } from "../../../../magmac/app/lang/JavaRules";
import { LazyRule } from "../../../../magmac/app/lang/LazyRule";
import { MutableLazyRule } from "../../../../magmac/app/lang/MutableLazyRule";
import { JavaEnumValues } from "../../../../magmac/app/lang/java/JavaEnumValues";
import { JavaLang } from "../../../../magmac/app/lang/java/JavaLang";
import { JavaMethod } from "../../../../magmac/app/lang/java/JavaMethod";
import { JavaStructureMember } from "../../../../magmac/app/lang/java/JavaStructureMember";
import { JavaStructureStatement } from "../../../../magmac/app/lang/java/JavaStructureStatement";
export class StructureMembers {
	deserialize(node : Node) : CompileResult<JavaStructureMember> {return 0.orError( 0, 0, 0.of( 0.wrap( 0.deserializeWhitespace), 0.wrap( 0.deserialize), 0.wrap( 0.deserialize), 0.wrap( 0.deserialize), 0.wrap( new JavaLang.JavaStructureNodeDeserializer( 0.JavaStructureType.Class)), 0.wrap( new JavaLang.JavaStructureNodeDeserializer( 0.JavaStructureType.Interface)), 0.wrap( new JavaLang.JavaStructureNodeDeserializer( 0.JavaStructureType.Record)), 0.wrap( new JavaLang.JavaStructureNodeDeserializer( 0.JavaStructureType.Enum))));;}
	createClassMemberRule() : Rule {break;break;break;break;break;return 0.set( new OrRule( 0.of( 0.createWhitespaceRule( ), 0.createStructureStatementRule( new TypeRule( 0, 0.createDefinitionRule( )), 0), 0.createMethodRule( 0), 0.createEnumValuesRule( 0), 0.createStructureRule( 0, 0), 0.createStructureRule( 0, 0), 0.createStructureRule( 0, 0), 0.createStructureRule( 0, 0))));;}
}
